File: backend/app/libs/api_resilience.py
"""API Resilience utilities for robust error handling and retry logic."""

import asyncio
import functools
import logging
import random
import time
from enum import Enum
from typing import Any, Callable, Optional, Type, Union, Dict, List
from fastapi import HTTPException
from dataclasses import dataclass


class ErrorSeverity(Enum):
    """Classification of error severity levels."""
    LOW = "low"           # User input errors, validation failures
    MEDIUM = "medium"     # Network timeouts, temporary service issues
    HIGH = "high"         # Database connection failures, external API failures
    CRITICAL = "critical" # Security errors, data corruption


class ErrorCategory(Enum):
    """Categories of errors for different handling strategies."""
    USER_ERROR = "user_error"           # 400-level errors caused by user input
    NETWORK_ERROR = "network_error"     # Network connectivity issues
    SERVICE_ERROR = "service_error"     # External service failures
    DATABASE_ERROR = "database_error"   # Database-related failures
    SYSTEM_ERROR = "system_error"       # Internal system errors
    SECURITY_ERROR = "security_error"   # Authentication/authorization errors


@dataclass
class RetryConfig:
    """Configuration for retry behavior."""
    max_attempts: int = 3
    base_delay: float = 1.0  # Base delay in seconds
    max_delay: float = 60.0  # Maximum delay in seconds
    exponential_backoff: bool = True
    jitter: bool = True  # Add randomness to prevent thundering herd
    retryable_errors: List[Type[Exception]] = None
    
    def __post_init__(self):
        if self.retryable_errors is None:
            self.retryable_errors = [
                ConnectionError,
                TimeoutError,
                OSError,  # Network-related errors
            ]


@dataclass
class ErrorContext:
    """Context information for error handling."""
    operation: str
    user_id: Optional[str] = None
    request_id: Optional[str] = None
    additional_data: Optional[Dict] = None


class ResilientError(Exception):
    """Enhanced error class with context and categorization."""
    
    def __init__(
        self,
        message: str,
        category: ErrorCategory = ErrorCategory.SYSTEM_ERROR,
        severity: ErrorSeverity = ErrorSeverity.MEDIUM,
        user_message: Optional[str] = None,
        suggestions: Optional[List[str]] = None,
        context: Optional[ErrorContext] = None,
        original_error: Optional[Exception] = None
    ):
        super().__init__(message)
        self.category = category
        self.severity = severity
        self.user_message = user_message or self._generate_user_message()
        self.suggestions = suggestions or []
        self.context = context
        self.original_error = original_error
        self.timestamp = time.time()
    
    def _generate_user_message(self) -> str:
        """Generate user-friendly error message based on category."""
        messages = {
            ErrorCategory.USER_ERROR: "Please check your input and try again.",
            ErrorCategory.NETWORK_ERROR: "Network connection issue. Please check your internet and try again.",
            ErrorCategory.SERVICE_ERROR: "External service temporarily unavailable. Please try again later.",
            ErrorCategory.DATABASE_ERROR: "Data access issue. Please try again in a moment.",
            ErrorCategory.SYSTEM_ERROR: "System error occurred. Our team has been notified.",
            ErrorCategory.SECURITY_ERROR: "Authentication required. Please log in and try again."
        }
        return messages.get(self.category, "An unexpected error occurred.")
    
    def to_http_exception(self) -> HTTPException:
        """Convert to FastAPI HTTPException with appropriate status code."""
        status_codes = {
            ErrorCategory.USER_ERROR: 400,
            ErrorCategory.NETWORK_ERROR: 503,
            ErrorCategory.SERVICE_ERROR: 502,
            ErrorCategory.DATABASE_ERROR: 503,
            ErrorCategory.SYSTEM_ERROR: 500,
            ErrorCategory.SECURITY_ERROR: 401
        }
        
        status_code = status_codes.get(self.category, 500)
        
        detail = {
            "message": self.user_message,
            "category": self.category.value,
            "severity": self.severity.value,
            "suggestions": self.suggestions
        }
        
        # Add context if available and not sensitive
        if self.context and self.category != ErrorCategory.SECURITY_ERROR:
            detail["operation"] = self.context.operation
            if self.context.request_id:
                detail["request_id"] = self.context.request_id
        
        return HTTPException(status_code=status_code, detail=detail)


def retry_with_backoff(config: RetryConfig = None):
    """Decorator for adding retry logic with exponential backoff."""
    if config is None:
        config = RetryConfig()
    
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            last_exception = None
            
            for attempt in range(config.max_attempts):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    
                    # Check if error is retryable
                    if not any(isinstance(e, error_type) for error_type in config.retryable_errors):
                        # If it's an HTTPException with 4xx status, don't retry
                        if isinstance(e, HTTPException) and 400 <= e.status_code < 500:
                            raise
                        # If it's a ResilientError with user error category, don't retry
                        if isinstance(e, ResilientError) and e.category == ErrorCategory.USER_ERROR:
                            raise
                        # For other non-retryable errors, convert and raise
                        if not isinstance(e, ResilientError):
                            resilient_error = ResilientError(
                                str(e),
                                category=ErrorCategory.SYSTEM_ERROR,
                                original_error=e
                            )
                            raise resilient_error.to_http_exception()
                        raise
                    
                    # Don't retry on last attempt
                    if attempt == config.max_attempts - 1:
                        break
                    
                    # Calculate delay
                    if config.exponential_backoff:
                        delay = min(config.base_delay * (2 ** attempt), config.max_delay)
                    else:
                        delay = config.base_delay
                    
                    # Add jitter to prevent thundering herd
                    if config.jitter:
                        delay *= (0.5 + random.random() * 0.5)  # 50-100% of calculated delay
                    
                    pass
                    await asyncio.sleep(delay)
            
            # If we get here, all retries failed
            if isinstance(last_exception, ResilientError):
                raise last_exception.to_http_exception()
            
            resilient_error = ResilientError(
                f"Operation failed after {config.max_attempts} attempts: {str(last_exception)}",
                category=ErrorCategory.SERVICE_ERROR,
                suggestions=["Please try again later", "Contact support if the issue persists"],
                original_error=last_exception
            )
            raise resilient_error.to_http_exception()
        
        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            last_exception = None
            
            for attempt in range(config.max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    
                    # Same retry logic as async version
                    if not any(isinstance(e, error_type) for error_type in config.retryable_errors):
                        if isinstance(e, HTTPException) and 400 <= e.status_code < 500:
                            raise
                        if isinstance(e, ResilientError) and e.category == ErrorCategory.USER_ERROR:
                            raise
                        if not isinstance(e, ResilientError):
                            resilient_error = ResilientError(
                                str(e),
                                category=ErrorCategory.SYSTEM_ERROR,
                                original_error=e
                            )
                            raise resilient_error.to_http_exception()
                        raise
                    
                    if attempt == config.max_attempts - 1:
                        break
                    
                    if config.exponential_backoff:
                        delay = min(config.base_delay * (2 ** attempt), config.max_delay)
                    else:
                        delay = config.base_delay
                    
                    if config.jitter:
                        delay *= (0.5 + random.random() * 0.5)
                    
                    pass
                    time.sleep(delay)
            
            if isinstance(last_exception, ResilientError):
                raise last_exception.to_http_exception()
            
            resilient_error = ResilientError(
                f"Operation failed after {config.max_attempts} attempts: {str(last_exception)}",
                category=ErrorCategory.SERVICE_ERROR,
                suggestions=["Please try again later", "Contact support if the issue persists"],
                original_error=last_exception
            )
            raise resilient_error.to_http_exception()
        
        # Return appropriate wrapper based on function type
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator


def handle_external_service_error(service_name: str, operation: str):
    """Decorator for handling external service errors (OpenAI, Stripe, etc.)."""
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            try:
                return await func(*args, **kwargs)
            except Exception as e:
                error_message = f"{service_name} {operation} failed: {str(e)}"
                
                # Categorize error based on the exception type and message
                if "rate limit" in str(e).lower() or "429" in str(e):
                    category = ErrorCategory.SERVICE_ERROR
                    suggestions = ["Please wait a moment and try again", "Consider upgrading your plan for higher limits"]
                elif "authentication" in str(e).lower() or "401" in str(e) or "403" in str(e):
                    category = ErrorCategory.SECURITY_ERROR
                    suggestions = ["Check your API credentials", "Ensure your account has the required permissions"]
                elif "timeout" in str(e).lower() or "connection" in str(e).lower():
                    category = ErrorCategory.NETWORK_ERROR
                    suggestions = ["Check your internet connection", "Try again in a few moments"]
                else:
                    category = ErrorCategory.SERVICE_ERROR
                    suggestions = [f"Try again later", f"Contact {service_name} support if the issue persists"]
                
                resilient_error = ResilientError(
                    error_message,
                    category=category,
                    severity=ErrorSeverity.MEDIUM,
                    suggestions=suggestions,
                    context=ErrorContext(operation=f"{service_name}:{operation}"),
                    original_error=e
                )
                
                raise resilient_error.to_http_exception()
        
        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except Exception as e:
                error_message = f"{service_name} {operation} failed: {str(e)}"
                
                if "rate limit" in str(e).lower() or "429" in str(e):
                    category = ErrorCategory.SERVICE_ERROR
                    suggestions = ["Please wait a moment and try again", "Consider upgrading your plan for higher limits"]
                elif "authentication" in str(e).lower() or "401" in str(e) or "403" in str(e):
                    category = ErrorCategory.SECURITY_ERROR
                    suggestions = ["Check your API credentials", "Ensure your account has the required permissions"]
                elif "timeout" in str(e).lower() or "connection" in str(e).lower():
                    category = ErrorCategory.NETWORK_ERROR
                    suggestions = ["Check your internet connection", "Try again in a few moments"]
                else:
                    category = ErrorCategory.SERVICE_ERROR
                    suggestions = [f"Try again later", f"Contact {service_name} support if the issue persists"]
                
                resilient_error = ResilientError(
                    error_message,
                    category=category,
                    severity=ErrorSeverity.MEDIUM,
                    suggestions=suggestions,
                    context=ErrorContext(operation=f"{service_name}:{operation}"),
                    original_error=e
                )
                
                raise resilient_error.to_http_exception()
        
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator


def validate_input(validation_fn: Callable, error_message: str = None):
    """Decorator for input validation with user-friendly errors."""
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            try:
                # Run validation
                validation_result = validation_fn(*args, **kwargs)
                if validation_result is False:
                    raise ValueError(error_message or "Invalid input provided")
                
                return func(*args, **kwargs)
            except ValueError as e:
                resilient_error = ResilientError(
                    str(e),
                    category=ErrorCategory.USER_ERROR,
                    severity=ErrorSeverity.LOW,
                    suggestions=["Please check your input and try again", "Ensure all required fields are filled correctly"]
                )
                raise resilient_error.to_http_exception()
        
        return wrapper
    return decorator


class CircuitBreaker:
    """Circuit breaker pattern implementation for external service calls."""
    
    def __init__(self, failure_threshold: int = 5, timeout: float = 60.0):
        self.failure_threshold = failure_threshold
        self.timeout = timeout
        self.failure_count = 0
        self.last_failure_time = None
        self.state = "closed"  # closed, open, half-open
    
    def can_execute(self) -> bool:
        """Check if the circuit breaker allows execution."""
        if self.state == "closed":
            return True
        elif self.state == "open":
            if time.time() - self.last_failure_time >= self.timeout:
                self.state = "half-open"
                return True
            return False
        else:  # half-open
            return True
    
    def record_success(self):
        """Record a successful execution."""
        self.failure_count = 0
        self.state = "closed"
    
    def record_failure(self):
        """Record a failed execution."""
        self.failure_count += 1
        self.last_failure_time = time.time()
        
        if self.failure_count >= self.failure_threshold:
            self.state = "open"
    
    def __call__(self, func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            if not self.can_execute():
                resilient_error = ResilientError(
                    "Service temporarily unavailable",
                    category=ErrorCategory.SERVICE_ERROR,
                    severity=ErrorSeverity.HIGH,
                    suggestions=["Please try again later", "The service will be available again shortly"]
                )
                raise resilient_error.to_http_exception()
            
            try:
                result = func(*args, **kwargs)
                self.record_success()
                return result
            except Exception as e:
                self.record_failure()
                raise
        
        return wrapper


# Common circuit breakers for external services
openai_circuit_breaker = CircuitBreaker(failure_threshold=3, timeout=30.0)
stripe_circuit_breaker = CircuitBreaker(failure_threshold=5, timeout=60.0)
firebase_circuit_breaker = CircuitBreaker(failure_threshold=3, timeout=30.0)

File: frontend/src/utils/api-resilience.ts
/**
 * Frontend API resilience utilities for robust error handling and retry logic.
 */

import { toast } from "sonner";

export enum ErrorSeverity {
  LOW = "low",
  MEDIUM = "medium", 
  HIGH = "high",
  CRITICAL = "critical"
}

export enum ErrorCategory {
  USER_ERROR = "user_error",
  NETWORK_ERROR = "network_error",
  SERVICE_ERROR = "service_error",
  SECURITY_ERROR = "security_error",
  SYSTEM_ERROR = "system_error"
}

export interface RetryConfig {
  maxAttempts: number;
  baseDelay: number; // milliseconds
  maxDelay: number; // milliseconds
  exponentialBackoff: boolean;
  jitter: boolean;
  retryableStatusCodes: number[];
}

export interface ErrorResponse {
  message: string;
  category: ErrorCategory;
  severity: ErrorSeverity;
  suggestions?: string[];
  operation?: string;
  requestId?: string;
}

export class ResilientError extends Error {
  public readonly category: ErrorCategory;
  public readonly severity: ErrorSeverity;
  public readonly userMessage: string;
  public readonly suggestions: string[];
  public readonly statusCode?: number;
  public readonly originalError?: Error;
  public readonly timestamp: number;

  constructor(
    message: string,
    category: ErrorCategory = ErrorCategory.SYSTEM_ERROR,
    severity: ErrorSeverity = ErrorSeverity.MEDIUM,
    userMessage?: string,
    suggestions: string[] = [],
    statusCode?: number,
    originalError?: Error
  ) {
    super(message);
    this.name = 'ResilientError';
    this.category = category;
    this.severity = severity;
    this.userMessage = userMessage || this.generateUserMessage();
    this.suggestions = suggestions;
    this.statusCode = statusCode;
    this.originalError = originalError;
    this.timestamp = Date.now();
  }

  private generateUserMessage(): string {
    const messages = {
      [ErrorCategory.USER_ERROR]: "Please check your input and try again.",
      [ErrorCategory.NETWORK_ERROR]: "Network connection issue. Please check your internet and try again.",
      [ErrorCategory.SERVICE_ERROR]: "Service temporarily unavailable. Please try again later.",
      [ErrorCategory.SECURITY_ERROR]: "Authentication required. Please log in and try again.",
      [ErrorCategory.SYSTEM_ERROR]: "An unexpected error occurred. Please try again."
    };
    return messages[this.category] || "An unexpected error occurred.";
  }

  showToast(): void {
    const toastFunctions = {
      [ErrorSeverity.LOW]: toast.info,
      [ErrorSeverity.MEDIUM]: toast.warning,
      [ErrorSeverity.HIGH]: toast.error,
      [ErrorSeverity.CRITICAL]: toast.error
    };

    const toastFn = toastFunctions[this.severity] || toast.error;
    toastFn(this.userMessage, {
      description: this.suggestions.length > 0 ? this.suggestions[0] : undefined,
      duration: this.severity === ErrorSeverity.CRITICAL ? 10000 : 5000
    });
  }
}

const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxAttempts: 3,
  baseDelay: 1000,
  maxDelay: 30000,
  exponentialBackoff: true,
  jitter: true,
  retryableStatusCodes: [408, 429, 500, 502, 503, 504]
};

export class CircuitBreaker {
  private failureCount = 0;
  private lastFailureTime = 0;
  private state: 'closed' | 'open' | 'half-open' = 'closed';

  constructor(
    private failureThreshold: number = 5,
    private timeout: number = 60000 // milliseconds
  ) {}

  canExecute(): boolean {
    if (this.state === 'closed') {
      return true;
    } else if (this.state === 'open') {
      if (Date.now() - this.lastFailureTime >= this.timeout) {
        this.state = 'half-open';
        return true;
      }
      return false;
    } else { // half-open
      return true;
    }
  }

  recordSuccess(): void {
    this.failureCount = 0;
    this.state = 'closed';
  }

  recordFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();
    
    if (this.failureCount >= this.failureThreshold) {
      this.state = 'open';
    }
  }

  getState(): string {
    return this.state;
  }
}

// Circuit breakers for different services
export const apiCircuitBreaker = new CircuitBreaker(3, 30000);
export const externalServiceCircuitBreaker = new CircuitBreaker(5, 60000);

export async function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  config: Partial<RetryConfig> = {},
  circuitBreaker?: CircuitBreaker
): Promise<T> {
  const retryConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: Error | null = null;

  // Check circuit breaker
  if (circuitBreaker && !circuitBreaker.canExecute()) {
    const error = new ResilientError(
      "Service temporarily unavailable",
      ErrorCategory.SERVICE_ERROR,
      ErrorSeverity.HIGH,
      "The service is currently unavailable. Please try again later.",
      ["Please wait a moment and try again", "Check the service status page"]
    );
    throw error;
  }

  for (let attempt = 0; attempt < retryConfig.maxAttempts; attempt++) {
    try {
      const result = await fn();
      
      // Record success in circuit breaker
      if (circuitBreaker) {
        circuitBreaker.recordSuccess();
      }
      
      return result;
    } catch (error) {
      lastError = error as Error;
      
      // Check if error is retryable
      const isRetryable = isRetryableError(error, retryConfig.retryableStatusCodes);
      
      if (!isRetryable || attempt === retryConfig.maxAttempts - 1) {
        // Record failure in circuit breaker
        if (circuitBreaker) {
          circuitBreaker.recordFailure();
        }
        
        // Convert to ResilientError if needed
        if (!(error instanceof ResilientError)) {
          throw convertToResilientError(error as Error);
        }
        throw error;
      }

      // Calculate delay for next attempt
      const delay = calculateDelay(attempt, retryConfig);
      console.log(`Retry attempt ${attempt + 1}/${retryConfig.maxAttempts} after ${delay}ms`);
      
      await sleep(delay);
    }
  }

  // This should never be reached, but just in case
  if (circuitBreaker) {
    circuitBreaker.recordFailure();
  }
  
  throw convertToResilientError(lastError!);
}

function isRetryableError(error: any, retryableStatusCodes: number[]): boolean {
  // Network errors are generally retryable
  if (error.name === 'TypeError' && error.message.includes('fetch')) {
    return true;
  }
  
  // Check status codes
  if (error.response?.status) {
    return retryableStatusCodes.includes(error.response.status);
  }
  
  // Check for specific error types
  if (error instanceof ResilientError) {
    return error.category === ErrorCategory.NETWORK_ERROR || 
           error.category === ErrorCategory.SERVICE_ERROR;
  }
  
  return false;
}

function calculateDelay(attempt: number, config: RetryConfig): number {
  let delay: number;
  
  if (config.exponentialBackoff) {
    delay = Math.min(config.baseDelay * Math.pow(2, attempt), config.maxDelay);
  } else {
    delay = config.baseDelay;
  }
  
  // Add jitter to prevent thundering herd
  if (config.jitter) {
    delay *= (0.5 + Math.random() * 0.5); // 50-100% of calculated delay
  }
  
  return Math.floor(delay);
}

function convertToResilientError(error: Error): ResilientError {
  // Try to parse error response for more context
  let category = ErrorCategory.SYSTEM_ERROR;
  let severity = ErrorSeverity.MEDIUM;
  let suggestions: string[] = [];
  let userMessage: string | undefined;
  let statusCode: number | undefined;

  // Check if it's a fetch error
  if (error.message.includes('fetch') || error.message.includes('network')) {
    category = ErrorCategory.NETWORK_ERROR;
    suggestions = ["Check your internet connection", "Try again in a moment"];
  }
  
  // Try to extract status code and details from response
  const errorString = error.toString();
  if (errorString.includes('400')) {
    category = ErrorCategory.USER_ERROR;
    severity = ErrorSeverity.LOW;
    statusCode = 400;
  } else if (errorString.includes('401') || errorString.includes('403')) {
    category = ErrorCategory.SECURITY_ERROR;
    severity = ErrorSeverity.MEDIUM;
    statusCode = parseInt(errorString.match(/\d{3}/)?.[0] || '401');
  } else if (errorString.includes('429')) {
    category = ErrorCategory.SERVICE_ERROR;
    severity = ErrorSeverity.MEDIUM;
    statusCode = 429;
    suggestions = ["Please wait a moment and try again", "You may be making requests too quickly"];
  } else if (errorString.includes('5')) {
    category = ErrorCategory.SERVICE_ERROR;
    severity = ErrorSeverity.HIGH;
    statusCode = parseInt(errorString.match(/5\d{2}/)?.[0] || '500');
  }

  return new ResilientError(
    error.message,
    category,
    severity,
    userMessage,
    suggestions,
    statusCode,
    error
  );
}

export function parseApiError(response: Response, responseData?: any): ResilientError {
  let category: ErrorCategory;
  let severity: ErrorSeverity;
  let suggestions: string[] = [];
  let userMessage: string;

  // Determine category and severity based on status code
  if (response.status >= 400 && response.status < 500) {
    if (response.status === 401 || response.status === 403) {
      category = ErrorCategory.SECURITY_ERROR;
      severity = ErrorSeverity.MEDIUM;
      suggestions = ["Please log in and try again", "Check your permissions"];
    } else if (response.status === 429) {
      category = ErrorCategory.SERVICE_ERROR;
      severity = ErrorSeverity.MEDIUM;
      suggestions = ["Please wait a moment and try again", "You may be making requests too quickly"];
    } else {
      category = ErrorCategory.USER_ERROR;
      severity = ErrorSeverity.LOW;
      suggestions = ["Please check your input and try again"];
    }
  } else if (response.status >= 500) {
    category = ErrorCategory.SERVICE_ERROR;
    severity = ErrorSeverity.HIGH;
    suggestions = ["Please try again later", "Contact support if the issue persists"];
  } else {
    category = ErrorCategory.SYSTEM_ERROR;
    severity = ErrorSeverity.MEDIUM;
  }

  // Extract user message from response
  if (responseData?.detail) {
    if (typeof responseData.detail === 'string') {
      userMessage = responseData.detail;
    } else if (responseData.detail.message) {
      userMessage = responseData.detail.message;
      if (responseData.detail.suggestions) {
        suggestions = responseData.detail.suggestions;
      }
    } else {
      userMessage = JSON.stringify(responseData.detail);
    }
  } else if (responseData?.message) {
    userMessage = responseData.message;
  } else {
    userMessage = `HTTP ${response.status}: ${response.statusText}`;
  }

  return new ResilientError(
    `API Error: ${userMessage}`,
    category,
    severity,
    userMessage,
    suggestions,
    response.status
  );
}

// Enhanced API call wrapper
export async function callApiWithResilience<T>(
  apiCall: () => Promise<Response>,
  config: Partial<RetryConfig> = {},
  circuitBreaker?: CircuitBreaker
): Promise<T> {
  return retryWithBackoff(async () => {
    const response = await apiCall();
    
    if (!response.ok) {
      let responseData;
      try {
        responseData = await response.json();
      } catch {
        // If we can't parse the response, use the status text
        responseData = { message: response.statusText };
      }
      
      const error = parseApiError(response, responseData);
      throw error;
    }
    
    return response.json() as Promise<T>;
  }, config, circuitBreaker);
}

// Helper function for showing user-friendly error messages
export function handleApiError(error: any, defaultMessage: string = "An unexpected error occurred"): void {
  if (error instanceof ResilientError) {
    error.showToast();
  } else {
    const resilientError = convertToResilientError(error);
    resilientError.showToast();
  }
  
  // Log the full error for debugging
  console.error('API Error:', error);
}

// Hook for API calls with automatic error handling
export function useResilientApi() {
  return {
    callApi: callApiWithResilience,
    handleError: handleApiError,
    retryWithBackoff,
    circuitBreaker: apiCircuitBreaker
  };
}
